src/tty/unix.rs

```diff
@@ -15,106 +15,76 @@
 //! tty related functionality
 //!
 
-use crate::tty::EventedReadWrite;
+use crate::tty::{EventedReadWrite, EventedPty, ChildEvent, ChildInfo};
 use crate::term::SizeInfo;
 use crate::display::OnResize;
 use crate::config::{Config, Shell};
 use crate::cli::Options;
 use mio;
 
-use libc::{self, c_int, pid_t, winsize, SIGCHLD, TIOCSCTTY, WNOHANG};
+use libc::{self, c_int, winsize, TIOCSCTTY};
+use nix::{ unistd::pipe, pty::openpty, sys::signal::* };
 
-use std::os::unix::io::{FromRawFd, RawFd};
+use std::os::unix::{ process::CommandExt, io::{FromRawFd, AsRawFd, RawFd} };
 use std::fs::File;
-use std::os::unix::process::CommandExt;
-use std::process::{Command, Stdio};
+use std::process::{Command, Stdio, Child};
 use std::ffi::CStr;
 use std::ptr;
 use mio::unix::EventedFd;
-use std::io;
-use std::os::unix::io::AsRawFd;
+use std::io::{self, Write, Read};
 
+/// Get the current value of `errno`
+fn errno() -> c_int {
+    ::errno::errno().0
+}
 
-/// Process ID of child process
+/// Set the current value of `errno`
 ///
-/// Necessary to put this in static storage for `sigchld` to have access
-pub static mut PID: pid_t = 0;
+/// This is needed to restore `errno` if it gets changed in a signal handler.
+fn set_errno(value: c_int) {
+    ::errno::set_errno(::errno::Errno(value));
+}
 
-/// Exit flag
+/// Self-pipe
 ///
-/// Calling exit() in the SIGCHLD handler sometimes causes opengl to deadlock,
-/// and the process hangs. Instead, this flag is set, and its status can be
-/// checked via `process_should_exit`.
-static mut SHOULD_EXIT: bool = false;
-
-extern "C" fn sigchld(_a: c_int) {
-    let mut status: c_int = 0;
-    unsafe {
-        let p = libc::waitpid(PID, &mut status, WNOHANG);
-        if p < 0 {
-            die!("Waiting for pid {} failed: {}\n", PID, errno());
-        }
-
-        if PID == p {
-            SHOULD_EXIT = true;
+/// We write and signals into the pipe, and the event loop wakes up and reads them.
+static mut SIG_PIPE_WRITE: Option<File> = None;
+
+extern "C" fn on_signal(sig: c_int) {
+    let old_errno = errno();
+
+    let byte = [sig as u8];
+    let file = unsafe { SIG_PIPE_WRITE.as_mut().unwrap() };
```

Comment:
While this is perfectly safe, I'm not sure if we want to take any chances here. It's an `Option` already anyways and we just unwrap that, but we could just make it something thread-safe and then return in case there's anything wrong.

An `Option<Arc<Mutex<File>>>` or something similar could just be checked without any potential for error at all, that would probably be a nice change.

Reply:
I'm not confident putting `Arc` or `Mutex` in there without first being certain they don't do anything that's not async-signal-safe. `Arc`, I assume, can `drop` and thus `free`? Not okay for signal handlers.

Honestly, even using `File` makes me slightly uncomfortable. I'm thinking about using a `RawFd` in an atomic and a `libc::write` instead.

Reply:
`Arc` or `Mutex` don't inherently do anything async unsafe. Quite the opposite they're used to *synchronize* access to a variable.

Now if the variable itself can't be accessed from multiple threads, that's a different thing. But then it shouldn't be publicly accessible at any time anyways. `Arc` or `Mutex` don't make this any worse, they just don't make it any better either.

Using a `RawFd` isn't necessarily any better either. And a `RawFd` can be wrapped in synchronization structs too.

The goal here is to make it impossible to do anything unsafe with it, so adding synchronization across threads is not going to make things worse.

Reply:
Apologies if I'm misreading you, but I think you may have misunderstood the issue here. Signal handlers aren't merely concurrent like threads, they also operate in a really strange, restrictive context, which makes various seemingly-innocuous operations suddenly really unsafe.

So far, `SIG_PIPE_WRITE` has been okay because it's set once _before_ the handler is installed, and then remains constant... but technically even then we're not 100% correct, because

1. We don't enforce any ordering between setting `SIG_PIPE_WRITE` and installing the handler, and
2. The access in `on_signal` isn't volatile.

There may be better solution to this, however; let me get back to you.

Reply:
> Signal handlers aren't merely concurrent like threads, they also operate in a really strange, restrictive context, which makes various seemingly-innocuous operations suddenly really unsafe.

I'm aware of this, however I wanted to point out that the current solution is also vulnerable to those issues. If we want to make things Rust-safe, using Arc and Mutex is an option in the global scope, but obviously an approach should be taken that is not only thread-safe, but also signal-safe.

Of course sometimes there's no way around adding an `unsafe` somewhere, however that should be carefully considered for potential abuse.